### PR TITLE
feat(metrics_extraction): Always include the environment in the query hash

### PR DESCRIPTION
**Pull request by armenzg**
Alerts and widgets are colliding since widgets only include the environment in the tags rather than in the query hash.

This change will require 90 days of baking and then we will switch customers over and drop the old spec version.

All widgets will double in number.

### CODE REVIEW

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 1]
```diff
@@ -1167,6 +1167,41 @@ def test_get_metrics_extraction_config_features_combinations(
...
+        "category": "transaction",
+        "condition": {"name": "event.duration", "op": "gte", "value": 10.0},
+        "field": None,
+        "mri": "c:transactions/on_demand@none",
+        "tags": tags,
+    }
+
+
+@django_db_all
+def test_include_environment_for_widgets(default_project: Project) -> None:
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1181]: Hi @k-fish 
Please have a close review of this test. I want to see what you would expect to see.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 2]
```diff
@@ -1167,6 +1167,41 @@ def test_get_metrics_extraction_config_features_combinations(
...
+    query = "transaction.duration:>=10"
+
+    with Feature([ON_DEMAND_METRICS, ON_DEMAND_METRICS_WIDGETS]):
+        create_widget(["count()"], query, default_project)
+        config = get_metric_extraction_config(default_project)
+        # Because we have two specs we will have two metrics.
+        # The second spec includes the environment tag as part of the query hash.
+        assert config and config["metrics"] == [
+            # _metric_spec([{"key": "query_hash", "value": "f1353b0f"}]),
+            _metric_spec([{"key": "query_hash", "value": "f1353b0f"}, env_tag]),
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1192]: Notice this env tag is included as part of the tags but it won't be the same down below.

File: src/sentry/snuba/metrics/extraction.py [hunk 3]
```diff
@@ -1179,11 +1180,16 @@ def mri(self) -> str:
     @cached_property
     def _query_str_for_hash(self) -> str:
         """Returns a hash of the query and field to be used as a unique identifier for the on-demand metric."""
-        str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()}"
+        if self.spec_version.flags == {"include_environment_tag"}:
+            str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()};{self._tag_for_field('environment')}"
+        else:
+            str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()}"
+
```
Comment (k-fish) [inline src/sentry/snuba/metrics/extraction.py:1187]: I feel like this should actually be in `self.groupbys` for the new version, and that should be a computed param based on the passed `groupbys` param to the spec, and optionally add env as a group-by. The new spec version can just have environment automatically added in the groupbys, which also means we can remove the condition in `to_metric_spec`.

Additionally we might want to add a check to avoid the case when you build a widget that already _has_ environment tag (since someone is explicitly grouping by env) in addition to the dynamic tag that's added, a group by doesn't need to be added twice.

File: tests/sentry/relay/config/test_metric_extraction.py (hunk 2 above)
Comment (k-fish) [reply tests/sentry/relay/config/test_metric_extraction.py:1260]: Yeah that's right, since widgets are `DYNAMIC_QUERY` and alerts are `SIMPLE_QUERY`

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 4]
```diff
@@ -1160,7 +1226,44 @@ def test_get_metrics_extraction_config_features_combinations(
...
+    }
+
+
+@django_db_all
+def test_include_environment_for_widgets(default_project: Project) -> None:
+    env_tag = {"field": "event.environment", "key": "environment"}
+    query = "transaction.duration:>=10"
+
+    with Feature([ON_DEMAND_METRICS, ON_DEMAND_METRICS_WIDGETS]):
+        create_widget(["count()"], query, default_project)
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1250]: @k-fish should this create_widget call include `{"field": "event.environment", "key": "environment"}` in the groupbys? (In other words, in the query hash)

What about if the create_widget call included columns with the environment?
Like this: `create_widget(["count()"], query, default_project, columns=["environment"])`?

File: src/sentry/snuba/metrics/extraction.py [hunk 5]
```diff
@@ -1136,6 +1137,15 @@ def __init__(
...
         self.groupbys = [groupby for groupby in groupbys or () if groupby != field]
+        # In the previous version we were not including the environment tag in the hash.
+        # Including it in the groupsby will include it in the query hash
+        if (
+            self.spec_type == MetricSpecType.DYNAMIC_QUERY
+            and "environment" not in self.groupbys
+            and self.spec_version.flags == {"include_environment_tag"}
+        ):
+            # XXX: Should this always be included?
+            self.groupbys.append("environment")
```
Comment (k-fish) [inline src/sentry/snuba/metrics/extraction.py:1148]: Yep, this is looking right, and yes, so long as it's `DYNAMIC_QUERY` environment should _always_ be there so people can change environment on the fly in their dashboards.

File: src/sentry/snuba/metrics/extraction.py [hunk 6]
```diff
@@ -1268,7 +1280,13 @@ def to_metric_spec(self, project: Project) -> MetricSpec:
...
 
-        if self.spec_type == MetricSpecType.DYNAMIC_QUERY:
+        # Once we switch to the next spec we can remove these two lines
+        # since the environment will be added to the groupbys, thus, being included in the query hash
+        if (
+            self.spec_type == MetricSpecType.DYNAMIC_QUERY
+            and self.spec_version.flags == {}
+            and self._tag_for_field("environment") not in extended_tags_conditions
+        ):
             extended_tags_conditions.append(self._tag_for_field("environment"))
```
Comment (k-fish) [inline src/sentry/snuba/metrics/extraction.py:1290]: Yep, cool 👍

**Review by k-fish — APPROVED** [review]
Looks good, once you get tests passing then 👍

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 7]
```diff
@@ -122,14 +122,17 @@ def test_get_metric_extraction_config_single_alert(default_project: Project) ->
         config = get_metric_extraction_config(default_project)
 
         assert config
-        assert len(config["metrics"]) == 1
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:125]: Some changes in this file are code-refactoring to make future spec version changes a little easier.

For instance, do not check the length of the metrics if we assert the contents of `config["metrics"]` right after.

I'm going to open another PR that will include these changes to make this PR easier to review.
You can ignore changes in this module until I get the PR in place.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 8]
```diff
@@ -122,14 +122,17 @@ def test_get_metric_extraction_config_single_alert(default_project: Project) ->
...
-        assert config["metrics"][0] == {
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:126]: I'm going to remove the usage of asserting specific indices of `config["metrics"]`.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 9]
```diff
@@ -1227,3 +1585,48 @@ def test_get_metric_extraction_config_with_no_spec(default_project: Project) ->
...
+        create_widget([aggr], query, default_project, columns=["environment"], title="foo")
+        config = get_metric_extraction_config(default_project)
+        assert config and config["metrics"] == [
+            _metric_spec([{"key": "query_hash", "value": "f1353b0f"}, env_tag]),
+            _metric_spec([{"key": "query_hash", "value": "4fb5a472"}, env_tag]),
+        ]
+
+        create_alert(aggr, query, default_project)
+        config = get_metric_extraction_config(default_project)
+        # XXX: Ask Kev about this
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1627]: @k-fish I think we should always prefer the widget over the alert. What do you think?

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 10]
```diff
@@ -1227,3 +1585,48 @@ def test_get_metric_extraction_config_with_no_spec(default_project: Project) ->
...
+    aggr = "count()"
+    env_tag = {"field": "event.environment", "key": "environment"}
+    query = "transaction.duration:>=10"
+
+    with Feature([ON_DEMAND_METRICS, ON_DEMAND_METRICS_WIDGETS]):
+        create_widget([aggr], query, default_project)
+        config = get_metric_extraction_config(default_project)
+        # Because we have two specs we will have two metrics.
+        # The second spec includes the environment tag as part of the query hash.
+        # XXX: Verify that the query hash is as expected
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1611]: I still need to do this.

I also need to add the code to query for a specific spec version based on an org's feature flag.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 11]
```diff
@@ -300,7 +300,17 @@ def test_get_metric_extraction_config_single_widget(default_project: Project) ->
                     {"key": "query_hash", "value": "a312e0db"},
                     {"key": "environment", "field": "event.environment"},
                 ],
-            }
+            },
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:303]: After merging #63947 the changes in this file are only about adding the new metric with a new hash (since the env is included in the calculation of the hash).

Please see the test at the bottom of this file which tests the change specifically.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 12]
```diff
@@ -1261,3 +1585,46 @@ def test_get_metric_extraction_config_with_no_spec(default_project: Project) ->
...
+        assert config and config["metrics"] == [
+            _metric_spec([{"key": "query_hash", "value": "f1353b0f"}, env_tag]),
+            _metric_spec([{"key": "query_hash", "value": "4fb5a472"}, env_tag]),
+        ]
+
+        create_alert(aggr, query, default_project)
+        config = get_metric_extraction_config(default_project)
+        # We loose the env in the tags because the alert spec is preferred over the widget spec
+        assert config and config["metrics"] == [
+            _metric_spec([{"key": "query_hash", "value": "f1353b0f"}]),
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1628]: @k-fish do you want me to fix that we pick up the spec from the alert instead of the widget one which includes the environment in the tags?

Should the alert spec include the env in the tags?

File: tests/sentry/search/events/builder/test_metrics.py [hunk 13]
```diff
@@ -2035,6 +2031,34 @@ def test_run_query_with_on_demand_count(self):
             ],
         )
 
+    # Once we delete the current spec version this test will fail and we can delete it
+    def test_on_demand_builder_with_new_spec(self):
+        field = "count()"
+        query = "transaction.duration:>0"
+        spec = OnDemandMetricSpec(field=field, query=query, spec_type=MetricSpecType.DYNAMIC_QUERY)
```
Comment (armenzg) [inline tests/sentry/search/events/builder/test_metrics.py:2038]: Even though, we do not pass a spec version with the latest spec version, when we use query builder with the feature flag it will pick up the latest spec version.

File: tests/sentry/search/events/builder/test_metrics.py [hunk 14]
```diff
@@ -2035,6 +2031,34 @@ def test_run_query_with_on_demand_count(self):
...
+        # As expected, it does not include the environment tag
+        expected_str_hash = "None;{'name': 'event.duration', 'op': 'gt', 'value': 0.0}"
+        assert spec._query_str_for_hash == expected_str_hash
```
Comment (armenzg) [inline tests/sentry/search/events/builder/test_metrics.py:2041]: Inside of the with block you will see that the query hash will include `["environment"]`.

File: tests/sentry/search/events/builder/test_metrics.py [hunk 15]
```diff
@@ -3062,7 +3086,6 @@ def test_run_query_with_on_demand_count_and_environments(self):
                     query=query_s,
                     environment=environment,
                     spec_type=MetricSpecType.SIMPLE_QUERY,
-                    spec_version=OnDemandMetricSpecVersioning.get_default_spec_version(),
```
Comment (armenzg) [inline tests/sentry/search/events/builder/test_metrics.py:3065]: This is unrelated to this PR. It's not necessary anymore.

File: tests/sentry/relay/config/test_metric_extraction.py (hunk 4 above)
Comment (k-fish) [reply tests/sentry/relay/config/test_metric_extraction.py]: > should this create_widget call include {"field": "event.environment", "key": "environment"} in the groupbys? (In other words, in the query hash)

Yes, every widget should always have `environment` in the group-by, which turns into `event.environment` in the metric spec. That saves environment as a tag until the metrics table, so when we query it later we can switch environment on the fly if a user changes env on their dashboard.

`create_widget(["count()"], query, default_project, columns=["environment"])`

Yeah the `columns` w/ `environment` means someone made a widget that explicitly grouped on environment, so we wouldn't repeat the tag.

File: tests/sentry/relay/config/test_metric_extraction.py (hunk 9 above)
Comment (k-fish) [reply tests/sentry/relay/config/test_metric_extraction.py]: hmm, a `widget` and `alert` with the new spec version should have different query hashes so they should no longer collide.

Can we remove the `columns=["environment"]` from the widget and test this for spec-version(1)? 

The result when testing an identical alert and widget (without an environment column in either) should be two separate query hashes, since the specs should be different. 

You'll have identical specs _except_ the widget spec will have an `environment` tag which is automatically added because of `DYNAMIC_QUERY`, and with spec version 1 that tag being there should also cause a different hash.

File: tests/sentry/relay/config/test_metric_extraction.py (hunk 12 above)
Comment (k-fish) [reply tests/sentry/relay/config/test_metric_extraction.py:1628]: Let's have both tests:
1. A test without environment in columns, they should be 2 different specs with 2 different hashes, no collision.
2. A test with environment in alert, but not in widget. This should make 2 identical specs with the same hash. Which one we pick here is irrelevant.

**Review by iambriccardo — APPROVED** [review]
I am lacking context of why we are adding the `environment` now, but functionally speaking the PR LGTM!

File: src/sentry/search/events/builder/metrics.py [hunk 16]
```diff
@@ -166,17 +166,13 @@ def _get_on_demand_metric_spec(self, field: str) -> Optional[OnDemandMetricSpec]
                     "Must include on demand metrics type when querying on demand"
                 )
 
-            # Instead of calling OnDemandMetricSpec without a spec_version (defaulting to the least),
-            # we keep this logic here to make future spec versions easier to roll out since it will
-            # save us to look where to add this logic again
-            spec_version = OnDemandMetricSpecVersioning.get_query_spec_version(self.organization_id)
-            return OnDemandMetricSpec(
+            return fetch_on_demand_metric_spec(
```
Comment (armenzg) [inline src/sentry/search/events/builder/metrics.py:169]: This change will make sure that the tests have the same function to get an `OnDemandMetricSpec` from a Relay metric spec.

File: src/sentry/snuba/metrics/extraction.py [hunk 17]
```diff
@@ -1430,6 +1444,27 @@ def _parse_query(value: str) -> QueryParsingResult:
             raise Exception(f"Invalid search query '{value}' in on demand spec: {e}")
 
 
+def fetch_on_demand_metric_spec(
```
Comment (armenzg) [inline src/sentry/snuba/metrics/extraction.py:1447]: This can now be used by the tests to have a single entry point to querying the right spec based on feature flags.

Jump to the end of `tests/sentry/relay/config/test_metric_extraction.py` to see.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 18]
```diff
@@ -1274,20 +1598,22 @@ def test_get_metric_extraction_config_with_no_spec(default_project: Project) ->
...
 
 
-# XXX: This may have bugs, therefore, use with caution
-def _on_demand_spec_from_widget(widget: DashboardWidgetQuery) -> OnDemandMetricSpec:
+def _on_demand_spec_from_widget(
+    project: Project, widget: DashboardWidgetQuery
+) -> OnDemandMetricSpec:
     field = widget.aggregates[0] if widget.aggregates else ""
-    return OnDemandMetricSpec(
+    return fetch_on_demand_metric_spec(
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1605]: I can now use this shared function to reduce differences between the code and the tests.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 19]
```diff
@@ -1310,33 +1636,40 @@ def _metric_spec(
...
+        assert spec._query_str_for_hash == expected_query_str_hash
+        assert spec.spec_version.version == 1
+        assert spec.spec_version.flags == set()
+
+        with Feature("organizations:on-demand-metrics-query-spec-version-two"):
+            spec = _on_demand_spec_from_widget(default_project, widget)
+            assert spec._query_str_for_hash == f"{expected_query_str_hash};['environment']"
+            assert spec.query_hash == "4fb5a472"
+            assert spec.spec_version.version == 2
+            assert spec.spec_version.flags == {"include_environment_tag"}
```
Comment (armenzg) [inline tests/sentry/relay/config/test_metric_extraction.py:1669]: Compare it without having the feature flag (just above):
```python
        spec = _on_demand_spec_from_widget(default_project, widget)
        assert spec.query_hash == "f1353b0f"
        assert spec._query_str_for_hash == expected_query_str_hash
        assert spec.spec_version.version == 1
        assert spec.spec_version.flags == set()
```

File: src/sentry/snuba/metrics/extraction.py [hunk 20]
```diff
@@ -80,12 +81,15 @@ class OnDemandMetricSpecVersioning:
...
     ]
 
     @classmethod
     def get_query_spec_version(cls: Any, organization_id: int) -> SpecVersion:
         """Return spec version based on feature flag enabled for an organization."""
-        _ = Organization.objects.get_from_cache(id=organization_id)
+        org = Organization.objects.get_from_cache(id=organization_id)
+        if features.has("organizations:on-demand-metrics-query-spec-version-two", org):
+            return cls.spec_versions[1]
         return cls.spec_versions[0]
```
Comment (k-fish) [inline src/sentry/snuba/metrics/extraction.py:94]: Not blocking, but doing this by array lookup is a little strange since that feature flag should _only_ be for SpecVersion(2), I think this can be hardcoded, which will make cleanup more obvious as well.

File: src/sentry/snuba/metrics/extraction.py [hunk 21]
```diff
@@ -1268,8 +1279,11 @@ def to_metric_spec(self, project: Project) -> MetricSpec:
         tag_from_groupbys = self.tags_groupbys(self.groupbys)
         extended_tags_conditions.extend(tag_from_groupbys)
 
+        # Once we switch to the next spec we can remove this block
+        # since the environment will be added to the groupbys, thus, being included in the query hash
         if (
             self.spec_type == MetricSpecType.DYNAMIC_QUERY
+            and self.spec_version.flags == set()
```
Comment (k-fish) [inline src/sentry/snuba/metrics/extraction.py:1286]: Not blocking; I don't know that we'll need this flag matching, it's pretty complicated, I can't see us needing a lot of individual behaviours and every new ability to a spec version will essentially reset the spec anyway.

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 22]
```diff
@@ -487,21 +591,21 @@ def test_get_metric_extraction_config_multiple_widgets_above_max_limit(
             config = get_metric_extraction_config(default_project)
             assert config
 
-            assert capture_exception.call_count == 1
+            assert capture_exception.call_count == 2
             exception = capture_exception.call_args.args[0]
             assert (
                 exception.args[0]
-                == "Spec version 1: Too many (2) on demand metric widgets for org baz"
+                == "Spec version 2: Too many (2) on demand metric widgets for org baz"
```
Comment (k-fish) [inline tests/sentry/relay/config/test_metric_extraction.py:598]: Okay so both spec version fail here and we're just checking the second one now?

File: tests/sentry/relay/config/test_metric_extraction.py [hunk 23]
```diff
@@ -1310,33 +1636,40 @@ def _metric_spec(
...
-        assert spec2._query_str_for_hash == f"{expected_query_str_hash};['environment']"
+        spec = _on_demand_spec_from_widget(default_project, widget)
+        assert spec.query_hash == "f1353b0f"
+        assert spec._query_str_for_hash == expected_query_str_hash
+        assert spec.spec_version.version == 1
+        assert spec.spec_version.flags == set()
+
+        with Feature("organizations:on-demand-metrics-query-spec-version-two"):
+            spec = _on_demand_spec_from_widget(default_project, widget)
+            assert spec._query_str_for_hash == f"{expected_query_str_hash};['environment']"
```
Comment (k-fish) [inline tests/sentry/relay/config/test_metric_extraction.py:1666]: This is specifically what we want for the new spec version, so 👍

**Review by k-fish — APPROVED** [review]
Looks good!

File: tests/sentry/relay/config/test_metric_extraction.py (hunk 22 above)
Comment (armenzg) [reply tests/sentry/relay/config/test_metric_extraction.py:598]: Both of them fail. I couldn't make up my mind. Perhaps I should not even take the version into consideration.

File: src/sentry/snuba/metrics/extraction.py (hunk 20 above)
Comment (armenzg) [reply src/sentry/snuba/metrics/extraction.py:94]: Yeah. I used to have a different approach that did not require any changes.
